src/qilib/utils/serialization.py

```diff
@@ -18,61 +18,184 @@
 OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 """
 
+from json import JSONEncoder, JSONDecoder
+from typing import Any, Dict, Callable, ClassVar
+from qilib.data_set import MongoDataSetIO
 import numpy as np
-import base64
-import pickle
-import serialize as serialize_module
 
+TransformFunction = Callable[[Any], Any]
 
-def __object_to_base64_encoded_pickle(u):
-    b = pickle.dumps(u)
-    return base64.b64encode(b).decode('ascii')
 
+class JsonSerializeKey:
+    """ The custum value types for the JSON serializer """
+    OBJECT = '__object__'
+    CONTENT = '__content__'
 
-def __base64_encoded_pickle_to_object(c):
-    b = base64.b64decode(c.encode('ascii'))
-    return pickle.loads(b)
 
+class Encoder(JSONEncoder):
+    """ A JSON encoder """
 
-serialize_module.register_class(np.ndarray, __object_to_base64_encoded_pickle, __base64_encoded_pickle_to_object)
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
 
+        self.encoders: ClassVar[Dict[type, TransformFunction]] = {}
 
-def _bytes_to_base64(b):
-    return base64.b64encode(b).decode('ascii')
+    def default(self, o):
+        if type(o) in self.encoders:
+            return self.encoders[type(o)](o)
 
+        return JSONEncoder.default(self, o)
 
-def _base64_to_bytes(c):
-    return base64.b64decode(c.encode('ascii'))
 
+class Decoder(JSONDecoder):
+    """ A JSON dncoder """
 
-serialize_module.register_class(bytes, _bytes_to_base64, _base64_to_bytes)
+    def __init__(self):
+        super().__init__(object_hook=self._object_hook)
 
-__msgfmt = 'json'
+        self.decoders: ClassVar[Dict[type, TransformFunction]] = {}
 
+    def _object_hook(self, obj):
+        if isinstance(obj, dict):
+            if JsonSerializeKey.OBJECT in obj:
+                if obj[JsonSerializeKey.OBJECT] in self.decoders:
+                    return self.decoders[obj[JsonSerializeKey.OBJECT]](obj)
+                else:
+                    raise ValueError()
 
-def serialize(data):
-    """ Serialize a Python object to JSON
+        return obj
 
-    Special objects might be serialized to a Python string.
+
+def _encode_bytes(data: Any) -> Dict[str, Any]:
+    return {JsonSerializeKey.OBJECT: bytes.__name__, JsonSerializeKey.CONTENT: data.decode('utf-8')}
+
+
+def _decode_bytes(data: Dict[str, Any]) -> bytes:
+    return data[JsonSerializeKey.CONTENT].encode('utf-8')
+
+
+class Serializer:
+    def __init__(self, encoders=None, decoders=None):
+        self.encoder = Encoder()
+        self.decoder = Decoder()
+
+        if encoders is None:
+            encoders = {}
+        self.encoder.encoders = encoders
+        if decoders is None:
+            decoders = {}
+        self.decoder.decoders = decoders
+
+        self.register_encoder(bytes, _encode_bytes)
+        self.register_decoder(bytes.__name__, _decode_bytes)
+        self.register_encoder(np.ndarray, MongoDataSetIO.encode_numpy_array)
```

Comment:
It is not so nice that the general `Serializer` used the encoder from `MongoDataSetIO` (especially since we will probably remove the `MongoDataSetIO` in the future.

Reply:
I think it's perfectly fine to reuse an encoder from another module. It's also used by `MongoDataSetIOReader` and qilib should work with numpy arrays. What is the point of having an extendible encode/decode system if we don't allow this?

Also, I don't think using a module/code that might or might not be removed in the unpredictable future is an argument to not use something what is already in production.